clang/lib/CConv/ConstraintVariables.cpp

```diff
@@ -1647,19 +1645,57 @@ void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV) {
   auto RetVar = getOnly(returnVars);
   RetVar->mergeDeclaration(FromRetVar);
   // Transplant params.
-  assert(From->numParams() == numParams());
-  for (unsigned i = 0; i < From->numParams(); i++) {
-    CVarSet &FromP = From->getParamVar(i);
-    CVarSet &P = getParamVar(i);
-    auto FromVar = getOnly(FromP);
-    auto Var = getOnly(P);
-    Var->mergeDeclaration(FromVar);
+  handle_params(From, this,
+                [](ConstraintVariable *F, ConstraintVariable *T)
+                { T->mergeDeclaration(F); });
+}
+
+void FunctionVariableConstraint::handle_params
+(FVConstraint *From, FVConstraint *To,
+ std::function<void(ConstraintVariable*,ConstraintVariable*)> f) {
+  bool fromEmpty = From->numParams() == 0;
+  bool toEmpty = To->numParams() == 0;
+  bool paramsEq = From->numParams() == To->numParams();
+  assert(fromEmpty || toEmpty || paramsEq);
+  if (paramsEq) {
+    for (unsigned i = 0; i < From->numParams(); i++) {
+      CVarSet &FromP = From->getParamVar(i);
+      CVarSet &P = To->getParamVar(i);
+      auto FromVar = getOnly(FromP);
+      auto Var = getOnly(P);
+      f(FromVar, Var);
+    }
+  } else { // Dealing with an untyped prototype
+    FVConstraint *Empty = fromEmpty ? From : To;
+    FVConstraint *Typed = fromEmpty ? To : From;
+    for (auto deferred : Empty->getDeferredParams()) {
+      assert(Typed->numParams() == deferred.PS.size());
+      for(unsigned i = 0; i < deferred.PS.size(); i++) {
+        CVarSet ParamDC = Typed->getParamVar(i);
+        CVarSet ArgDC = deferred.PS[i];
+        ProgramInfo &Info = deferred.I;
+        auto &CS = Info.getConstraints();
+        constrainConsVarGeq(ParamDC, ArgDC, CS, &(deferred.PL), Wild_to_Safe, false, &Info);
```

Comment:
It seems that what you are missing is all of the stuff involving array bounds. Why is it not here?

Reply:
It relies on a good deal of context. For example, it requires still having access to the Argument objects. Will these still be in memory or do they get free'd?

Reply:
I don't actually know. @Machiry do you know?

Reply:
@aaronjeline No.  After recent refactoring, we no more have handles to AST objects. We use `BoundsKey` as an identifier now. Look at `constrainLocalAssign` and `constrainConsVarGeq`:
```
        // Add assignment to bounds info graph.
        if (PCLHS->hasBoundsKey() && PCRHS->hasBoundsKey()) {
          Info->getABoundsInfo().addAssignment(PCLHS->getBoundsKey(),
                                               PCRHS->getBoundsKey());
        }
```